include/ginkgo/core/reorder/scaled_reordered.hpp

```diff
@@ -0,0 +1,257 @@
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_REORDER_SCALED_REORDERED_HPP_
+#define GKO_PUBLIC_CORE_REORDER_SCALED_REORDERED_HPP_
+
+
+#include <ginkgo/core/base/abstract_factory.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+#include <ginkgo/core/matrix/identity.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/reorder/reordering_base.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace reorder {
+
+
+/**
+ * Provides an interface to wrap reorderings like Rcm and diagonal scaling
+ * like equilibration around a LinOp like e.g. a sparse direct solver.
+ *
+ * Reorderings can be useful for reducing fill-in in the numerical factorization
+ * phase of direct solvers, diagonal scaling can help improve the numerical
+ * stability by reducing the condition number of the system matrix.
+ *
+ * With a permutation matrix P, a row scaling R and a column scaling C, the
+ * inner operator is applied to the system matrix P*R*A*C*P^T instead of A.
+ * Instead of A*x = b, the inner operator attempts to solve the equivalent
+ * linear system P*R*A*C*P^T*y = P*R*b and retrieves the solution x = C*P^T*y.
+ * Note: The inner system matrix is computed from a clone of A, so the original
+ * system matrix is not changed.
+ *
+ * @tparam ValueType  Type of the values of all matrices used in this class
+ * @tparam IndexType  Type of the indices of all matrices used in this class
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class ScaledReordered
+    : public EnableLinOp<ScaledReordered<ValueType, IndexType>> {
+    friend class EnableLinOp<ScaledReordered, LinOp>;
+    friend class EnablePolymorphicObject<ScaledReordered, LinOp>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using ReorderingBaseFactory =
+        AbstractFactory<gko::reorder::ReorderingBase<IndexType>,
+                        gko::reorder::ReorderingBaseArgs>;
+
+    std::shared_ptr<const LinOp> get_system_matrix() const
+    {
+        return system_matrix_;
+    }
+
+    std::shared_ptr<const LinOp> get_inner_operator() const
+    {
+        return inner_operator_;
+    }
+
+    GKO_CREATE_FACTORY_PARAMETERS(parameters, Factory)
+    {
+        /**
+         * The inner operator factory that is to be generated on the scaled
+         * and reordered system matrix.
+         */
+        std::shared_ptr<const LinOpFactory> GKO_FACTORY_PARAMETER_SCALAR(
+            inner_operator, nullptr);
+
+        /**
+         * The reordering that is to be applied to the system matrix.
+         */
+        std::shared_ptr<const ReorderingBaseFactory>
+            GKO_FACTORY_PARAMETER_SCALAR(reordering, nullptr);
```

Comment:
Requires a note that the system matrix must be of type `Permutable<index_type>` if a reordering is specified.